### PR TITLE
LSTM support for ClippedPPO

### PR DESCRIPTION
**Pull request by shadiendrawis**

### CODE REVIEW

File: rl_coach/agents/clipped_ppo_agent.py [hunk 1]
```diff
@@ -151,55 +151,109 @@ def set_session(self, sess):
...
+                for i in range(0, episode.length()-sequence_length, stride):
+                    trajectory = self.pre_network_filter.filter(episode.transitions[i:i+sequence_length],
+                                                                deep_copy=False,
+                                                                update_internal_state=update_internal_state)
+                    transition = Transition()
+                    transition.add_info({'sample_sequence': Batch(trajectory)})
+                    dataset.append(transition)
+        else:
+            dataset = self.memory.transitions
+            update_internal_state = self.ap.algorithm.update_pre_network_filters_state_on_train
```
Comment (gal-leibovich) [inline rl_coach/agents/clipped_ppo_agent.py:171]: already done in line 156

File: rl_coach/agents/clipped_ppo_agent.py [hunk 2]
```diff
@@ -314,22 +386,12 @@ def train(self):
...
-            update_internal_state = self.ap.algorithm.update_pre_network_filters_state_on_train
-            dataset = self.pre_network_filter.filter(dataset, deep_copy=False,
-                                                     update_internal_state=update_internal_state)
-            batch = Batch(dataset)
-            for training_step in range(self.ap.algorithm.num_consecutive_training_steps):
-                self.networks['main'].sync()
-                self.fill_advantages(batch)
-
-                # take only the requested number of steps
-                if isinstance(self.ap.algorithm.num_consecutive_playing_steps, EnvironmentSteps):
```
Comment (gal-leibovich) [inline rl_coach/agents/clipped_ppo_agent.py:327]: where did the `take only the requested number of steps` and the shuffling go? we don't need shuffling with the LSTM sequences, but how about when not using LSTM?

File: rl_coach/architectures/tensorflow_components/middlewares/lstm_middleware.py [hunk 3]
```diff
@@ -57,20 +59,25 @@ def _build_module(self):
...
         self.c_init = np.zeros((1, lstm_cell.state_size.c), np.float32)
         self.h_init = np.zeros((1, lstm_cell.state_size.h), np.float32)
-        self.state_init = [self.c_init, self.h_init]
+        if self.sequence_length == -1:
+            rnn_in = tf.expand_dims(self.layers[-1], [0])
+        else:
+            seq_len = tf.cond(tf.equal(tf.shape(self.layers[-1])[0], tf.constant(1)),
+                              lambda: 1, lambda: self.sequence_length)
+            rnn_in = tf.reshape(self.layers[-1], [-1, seq_len] + list(self.layers[-1].shape[1:]))
+        batch_size = tf.shape(rnn_in)[0]
```
Comment (shadiendrawis) [inline rl_coach/architectures/tensorflow_components/middlewares/lstm_middleware.py:72]: combine both cases into one, and document how this is supposed to work.

File: rl_coach/architectures/tensorflow_components/middlewares/lstm_middleware.py [hunk 4]
```diff
@@ -57,20 +59,25 @@ def _build_module(self):
             ))
 
         # add the LSTM layer
-        lstm_cell = tf.nn.rnn_cell.BasicLSTMCell(self.number_of_lstm_cells, state_is_tuple=True)
+        lstm_cell = tf.nn.rnn_cell.LSTMCell(self.number_of_lstm_cells, state_is_tuple=True,
+                                           activation=self.activation_function)
         self.c_init = np.zeros((1, lstm_cell.state_size.c), np.float32)
         self.h_init = np.zeros((1, lstm_cell.state_size.h), np.float32)
```
Comment (shadiendrawis) [inline rl_coach/architectures/tensorflow_components/middlewares/lstm_middleware.py:65]: documentation on how resetting to initial state works.

File: rl_coach/agents/clipped_ppo_agent.py [hunk 5]
```diff
@@ -151,55 +151,109 @@ def set_session(self, sess):
...
+            else:
+                feed_dict[k] = data.states([k])[k][start:end]
+        return feed_dict
+
     def fill_advantages(self, batch):
         network_keys = self.ap.network_wrappers['main'].input_embedders_parameters.keys()
 
         state_values = []
         for i in range(int(batch.size / self.ap.network_wrappers['main'].batch_size) + 1):
+            self.networks['main'].online_network.reset_internal_memory()
```
Comment (shadiendrawis) [inline rl_coach/agents/clipped_ppo_agent.py:190]: rename function
